### DISCUSSION TITLE
Using V-HACD to generate convex hull shapes for bodies.

### DISCUSSION THREAD

**Discussion by Pindrought** [jrouwe/JoltPhysics#726] [Q&A] [answered]
I have this basic cottage model that I found online to use for testing.
![image](https://github.com/jrouwe/JoltPhysics/assets/39708313/f29f896c-bc6a-466a-86d2-e0fcced22af2)

I am trying out using V-HACD with different parameters to generate a combination of convex hull shapes to represent the physics body for this entity.

One issue I am running into is that to get something close to the shape of the cottage, i'm having to use a lot of convex hull shapes. 

Here is an example with 64 shapes.
![image](https://github.com/jrouwe/JoltPhysics/assets/39708313/c1b9f2cb-9690-478a-b323-be6ed202ca67)

Here is an example with 128 shapes.
![image](https://github.com/jrouwe/JoltPhysics/assets/39708313/e5a08ddc-b827-4ace-8152-57cef641dbc2)

I know that I could manually determine the shapes and get it to a much smaller amount, but it's not realistic for me to manually do this for every model.

Is this to be expected for a model like this? Should I be approaching this a different way? Is a physics body made up of 128 convex hull shapes going to perform poorly? Should I ever be using MeshShapes or should that always be a last resort? Is the real solution here to have a 3d artist specifically for building separate meshes used for collider generation?

Thanks!

**Comment by jrouwe** [comment]
I don't know enough about V-HACD to say if this is is normal or not, perhaps ask the question at the [V-HACD](https://github.com/kmammou/v-hacd) repo?

It looks like this house is not going to move though: a ```MeshShape``` should be fine for any body that is ```EMotionType::Static``` (and probably faster than a compound of 128 convex hulls). Only dynamic or kinematic objects need to be converted into a (compound of) convex hull(s).

**Reply by Pindrought** [reply]
Unfortunately the author of V-HACD is in the process of moving and doesn't look like they'll be able to assist in the near future. I had made another post about an issue where meshes with 0 or near 0 volume are not processed properly which makes sense since V-HACD uses voxels/volume to build the shapes.

It is good to know that mesh shapes are fine for static bodies. Is this true regardless of the complexity of the mesh?

I am also going to dive into trying out this alternative convex decomposition library and see if it gives more favorable results. https://colin97.github.io/CoACD/

Do you know for Horizon Forbidden West, are mesh shapes usually used for the static objects or is it a combination of things?

**Reply by jrouwe** [reply, marked as answer]
> Is this true regardless of the complexity of the mesh?

Your building looks quite low poly so will work fine. If you have high poly meshes you may want to push your mesh through a mesh optimizer like Simplygon (don't know if there's a good open source alternative). In general you want your dynamic objects to hit ~10 triangles at the same time (so centimeter sized triangles won't work well with meter sized dynamic objects as you would hit 100 triangles which can cause the simulation to slow down).

> Do you know for Horizon Forbidden West, are mesh shapes usually used for the static objects or is it a combination of things?

Almost everything static is a MeshShape standing on a HeightFieldShape.
